### PR TITLE
feat: add support for container literals and in operator in ExprWrapper

### PR DESCRIPTION
**Pull request by Raine-Yang-UofT**
## Proposed Changes

Extend `ExprWrapper` to support `List/Set/Tuple` literals and `in/not in` operators.
...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |     X     |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: tests/test_z3_visitor.py [hunk 1]
```diff
@@ -54,3 +54,33 @@ def f(x: bool, y: bool, z: bool, a):
         solver = z3.Solver()
         solver.add(e == a)
         assert solver.check() == z3.sat
+
+
+def test_container_constraints():
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply tests/test_z3_visitor.py:66]: This is better, but can still be improved by taking advantage of pytest's test parametrization. Check out some of the past PRs from your teammates if you want to see some examples of this in action.

[the remark this answers is not in the document]
Comment (Raine-Yang-UofT) [reply tests/test_z3_visitor.py:66]: Hi Professor, sorry that I didn't fully understanding how you want me to use pytest's test parameterization. Do we want to only parameterize the test for `list/set/tuple` or all the tests in the module. Currently I'm using test parameterization inside `test_container_constraints` function, and I will try to incorporate more code into the inner `check_constraint` function.

One difficulty for parameterizing all test cases is that we cannot duplicate the z3 variables we define. For the tests we need to define z3 variables such as `x = z3.Int("x")` for the expected z3 constraints. If we want to move the list of code and expected constraints outside the functions these variables will become global, so we cannot reuse a variable for a different type once its type is defined. This is also the reason why I'm keeping the parameterized test within the outer `test_container_constraints` to limit the scope of variables.

File: tests/test_z3_visitor.py [hunk 2]
```diff
@@ -41,16 +30,127 @@ def f(x: bool, y: bool, z: bool, a):
...
-        solver.add(e == a)
-        assert solver.check() == z3.sat
+    return function_def.z3_constraints
+
+
+@pytest.mark.parametrize("code, expected", zip(code_list, expected_list))
+def test_constraint(code, expected):
+    for function, function_expected in zip(code, expected):
+        function_actual = _get_constraints_from_code(function)
+        assert len(function_actual) == len(function_expected)
```
Comment (Raine-Yang-UofT) [inline tests/test_z3_visitor.py:152]: Although checking the length of expected and actual expressions to be equal has been our existing approach, and is used to ensure `zip` function works correctly, I'm not sure this is the best practice to perform the test. In principle, we do not need the number of expressions to be equal to that asseted, or the expressions appear in the exact same order, as long as the two list of expressions are equivalent. Our approach seems to be testing against our specifc implementation instead of the module's feature.

Comment (david-yz-liu) [reply tests/test_z3_visitor.py:152]: While the relevant module functionality is to convert the preconditions into Z3 expressions, it is appropriate for the module to do this conversion as a one-to-one mapping (i.e., without further simplification) until we decide to change this.

More to the point, however, this reveals a limitation on our current testing approach. The changes you made were to `ExprWrapper`, which currently is only tested indirectly through the tests in this file for `Z3Visitor`. Certainly I would not accept any simplification in this file until there were unit tests for `ExprWrapper` itself. (I'm not saying to add those in this PR, but rather to explain why _even if I approve simplification of the precondition expressions_ generally, I would not approve changing the tests in this file without additional tests added.)

File: tests/test_z3_visitor.py [hunk 3]
```diff
@@ -41,16 +30,127 @@ def f(x: bool, y: bool, z: bool, a):
...
+# expected container expressions
+x = z3.Int("x")
+container_expected = [
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.BoolVal(False), z3.BoolVal(False)],
```
Comment (david-yz-liu) [inline tests/test_z3_visitor.py:129]: First, each of these should be in their own function (as you've done for all of the above test cases in `container_expected`). Same with the "not in empty" test below.

Second (more as a comment), I'm surprised that there are three precondition expressions but only two Z3 expressions here. Can you explain why that is?

Comment (Raine-Yang-UofT) [reply tests/test_z3_visitor.py:129]: Hi Professor, sorry that I wan't quite clear what do you mean by the first comment. I didn't see how the test cases for empty containers differ from the ones above.

About the second problem, sorry that I wasn't being careful enough. The precondition for empty set turns out to be ignored since PythonTA couldn't infer the return type of `set()` function and skipped creating z3 expression for this precondition. However, I cannot fix this problem with modifications in `ExprWrapper` as the problem lies in the `parse_assertions` function in `contracts/__init__.py`.  Currently we may not be able to create the test cases with empty sets.

Comment (david-yz-liu) [reply tests/test_z3_visitor.py:129]: @Raine-Yang-UofT sure, what I meant was that I'd like to see each precondition as a separate test case, i.e., splitting up the `x in []` and `x in ()` preconditions into separate test cases (and function). This aligns with the other test cases you've added, which each have one precondition per function.

Good point about the usage of `set()`. It's fine to not include test cases for empty sets in this PR.

Comment (Raine-Yang-UofT) [reply tests/test_z3_visitor.py]: @david-yz-liu Thanks, I have updated the tests. In addition, perhaps I can submit a issue concerning the problem with empty sets.

**Review by david-yz-liu — APPROVED** [review]
